### ISSUE TITLE
float16 precision might lead to initial_rgb's initialization with NaN tensor

### ISSUE THREAD

**Issue by sandmartspinoff** [princeton-computational-imaging/NeuLS#6]
Hello, I am very interested in your work and was testing the tutorial (train.py in /NeuLS repository) on my jupyter notebook environment.
And I realized that rgb output from the light sphere model's forward function is NaN tensor.

![Image](https://github.com/user-attachments/assets/5c28e014-7634-4ffd-b4c3-154ea016fa8b)

I guess it's from the initialization of "initial_rgb" 's data in line 623 of train.py

![Image](https://github.com/user-attachments/assets/c410ba59-722e-42a6-9a60-e4e7927480f6)

where "model.data.rgb_volume[0]" has type "torch.float16".
I think error occurs from the operation of torch.mean with type "torch.float16"

Fortunately, it seems that simply converting to float32 works as follows:

![Image](https://github.com/user-attachments/assets/f5e83a0f-e6b7-4031-98ec-bdb9080a35ab)

It seems to work by changing the line 623

![Image](https://github.com/user-attachments/assets/da005a0c-a7c1-4f4d-aa8c-47a788c0293b)

**Comment by Ilya-Muromets** [comment]
Interesting, just to confirm this was your own jupyter notebook env (i.e. not Colab)?
If so, what version of the packages is this with?

**Comment by sandmartspinoff** [comment]
Yes, I'm not using Colab, I'm using a docker environment from the base image of python:3.10 and
version of the packages are:

torch==2.0.1+cu118
torchvision==0.15.2+cu118
pytorch_lightning==2.0.1

numpy==1.26.4
matplotlib==3.7.0
and
tinycudann installed from .whl file downloaded from the link supplied in the tutorial of NSF (Neural spline field) :
gdown "https://drive.google.com/u/1/uc?id=1-7x7qQfB7bIw2zV4Lr6-yhvMpjXC84Q5&confirm=t"
pip install tinycudann-1.7-cp310-cp310-linux_x86_64.whl

I tried with tinycudann with the link supplied in the tutorial of NeuLS, but it didn't work so,,

**Comment by sandmartspinoff** [comment]
Moreover, I have some questions about your model.

1. I understood FOV Scale as a parameter for zooming out and zooming in the current time frame's scene. You introduced two ways of doing this as:

- render_image function of tutorial rescales uv coordinates as:

`adjusted_uv = uv * fov_scale + 0.5 * (1 - fov_scale)  # re-scale coordinate to match FOV`

- generate_output function of train.py multiplies fov_scale to the intrinsics_inv matrix's first row
```
intrinsics_inv = intrinsics_inv.clone()
intrinsics_inv[:,0] = intrinsics_inv[:,0] * fov_scale
intrinsics_inv[:,1] = intrinsics_inv[:,1]
```

I guess you did them for the same purpose, but I don't understand how fov_scale is used in 2). (why fov_scale is multiplied to the first row of the matrix, not the whole matrix) 


2. I think you multiplied some factors to make the resultant translation vector of each frame's camera within the sphere with radius 1, and I like to ask what meaning focal_compensation, and the value 1.9236 have in the followings of train.py:

`return self.args.focal_compensation * self.args.translation_weight * (translation + 5 * self.center)`

```
if training_phase < 0.2: # Apply random perturbation for training during early epochs
            factor = 0.015 * self.args.focal_compensation
            perturbation = torch.randn_like(ray_origins) * factor / 0.2 * max(0.0, 0.25 - training_phase)
            ray_origins = ray_origins + perturbation

```

3. Lastly, I want to know the reason why you encoded "middle point of intersection point P and (1,1,1)" not simply P in the light sphere model:

`encoded_offset_position = self.encoding_offset_position((intersections_sphere + 1) / 2)`

`encoded_color_position = self.encoding_color_position((intersections_sphere_offset + 1) / 2)`

Thank you :)

**Comment by Ilya-Muromets** [comment]
Interesting, yeah I did not run into that particular NaN issue on my machine, but will leave this up as a solution to anyone who does.

Regarding the followup questions:
1. I (arbitrarily) expand the FOV only horizontally (e.g., going from portrait mode to wide images), hence multiplying only the first row of the intrinsics matrix (corresponding to fx and cx). You can also multiply the second row to expand vertically.

2.  If you see the comment on line 388 of train.py: https://github.com/princeton-computational-imaging/NeuLS/blob/6128a57479d9edaf4675df5c8660e471f1d57e94/train.py#L389C9-L389C109

This basically normalizes motion/perturbations to the scale of the 82deg Pixel 8 Pro main lens (that's the 1.9236 factor (fx/cx)). Essentially, if you have super zoomed in telephoto data, we don't want to magnify the perturbations on the rays (as a percentage of the effective FOV). 

3. Neural field encodings want values you pass into them to be between 0 and 1, but our unit sphere is between -1 and 1, so I'd think of this less as "middle point of intersection point P and (1,1,1)", but more as P normalized to 0-1.

**Comment by sandmartspinoff** [comment]
Thank you very much!

**Comment by sandmartspinoff** [comment]
Hello :) , it's been a while, but I still don't clearly understand what ray offset model do. 

In the context of light field, I thought you try to parameterize rays that go outside the sphere with two intersection points (uv of image frame and P of sphere intersection). However, in this perspective, this parametrization cannot fully express all such rays. (since different rays that come with different direction while intersecting same point P on the sphere can be from same uv coordinates). 

But I thought this would be fine for panoramic capturing settings since there's not much disparity differences for same objects across different image frames.

And just as a brief understanding, I thought ray offset model kind of extends the expressiveness of the model than that of single view dependent color model. But It's kind of hard for me to think of how the idea of ray offset specifically deals with what specific cases.

Can you explain a bit more about this?

And I also tried your model with the panoramic captured frames with a fastly moving robot, and like as you have mentioned in the paper with the result of "CityCars" dataset, in the early epochs it reconstructs some small blurry object then after many epochs it reconstructs moving robots across frames. And also, I found "Fireworks" dataset in your repository. Was this also your intention to enable the reconstruction of moving objects along capturing time.

**Comment by Ilya-Muromets** [comment]
Heya, I sent you an email if you're interested in having a zoom meeting over this.

the TLDR is basically yes, you can have multiple rays from the same UV and intersection point on the sphere that are distinct (e.g., different rotation or time). The light sphere lumigraph-ey model I have is purposely underparametrized as compared to a full lightfield, as we really don't get a lot of overlapping views from a panorama capture so don't have a lot of observations to fit some high dimensional light field function.

You can think, in simplest terms, of the ray offset model to be a sort of a weird optical flow on the surface of the sphere. An example of this thatcan be helpful to draw out is how you can model parallax on the sphere as this ray bending (e.g., I move my camera, the intersections on the sphere move a little, and the ray offset model can multiply that motion to "compensate" for the actual parallax of that point in the scene). 

Yes, CityCars and Fireworks are intended to be mostly failure cases (but maybe you can make them work!). Fireworks is even worse than large motion, as the colors from the explosion are spontaneously generated (i.e. you can't just move the colors from one frame to another to reconstruct it). Unlike the cars, where if you had a semi-static car reconstruction, you can then rigidly move it across the screen to simulate the recorded video.

**Comment by Ilya-Muromets** [comment]
btw I updated the tinycudann wheels in NeuLS/NSF so the colab should work now I think? I guess I basically need to update these every time colab gets updated

**Comment by sandmartspinoff** [comment]
Yes, It's working now!
